blaze/compute/sql.py

```diff
@@ -79,25 +80,77 @@ def inner_columns(s):
     raise TypeError()
 
 
-@dispatch(Projection, Selectable)
+def ensure_bind(f):
+    """Ensure that the bind is properly copied up after compute_up.
+
+    Parameters
+    ----------
+    f : callable
+        The compute_up function to wrap.
+
+    Returns
+    -------
+    wrapped : callable
+        The wrapped compute_up.
+    """
+    @wraps(f)
+    def wrapped(expr, *args, **kwargs):
+        if not args:
+            return f(expr, *args, **kwargs)
+
+        data, args = args[0], args[1:]
+        if isinstance(data, sa.engine.Engine):
+            bind = data
+        else:
+            bind = data.bind
+        e = f(expr, data, *args, **kwargs)
+        try:
+            e.bind = bind
+        except AttributeError:
+            pass
+        return e
+
+    return wrapped
+
+
+def dispatch_and_bind(*args):
+    """Custom dispatch that wraps the function in an ``ensure_bind``.
+
+    Parameters
+    ----------
+    *args
+        Forwarded to ``blaze.dispatch.dispatch``
+    **kwargs
+        Forwarded to ``blaze.dispatch.dispatch``
+
+    Returns
+    -------
+    dispatcher : callable
+        A decorator that will both wrap the function in ``ensure_types``
+        and the normal ``blaze.dispatch.dispatch``.
+    """
+    return toolz.compose(dispatch(*args), ensure_bind)
+
+
+@dispatch_and_bind(Projection, Selectable)
```

Comment:
I was just trying to prevent people from forgetting to use this new decorator in the future. `binop_sql` uses the two decorator aproach. If you think this is just more confusing, I can change it.